src/Server/HTTPHandlerFactory.cpp

```diff
@@ -74,85 +74,59 @@ static inline auto createHandlersFactoryFromConfig(IServer & server, const std::
+            else if (handler_type == "predefined_query_handler")
+                main_handler_factory->addHandler(createPredefinedHandlerFactory(server, prefix + "." + key));
+            else if (handler_type == "prometheus")
+                main_handler_factory->addHandler(createPrometheusHandlerFactory(server, async_metrics, prefix + "." + key));
+            else if (handler_type == "replicas_status")
+                main_handler_factory->addHandler(createReplicasStatusHandlerFactory(server, prefix + "." + key));
+            else
+                throw Exception("Unknown handler type '" + handler_type + "' in config here: " + prefix + "." + key + ".handler.type",
+                    ErrorCodes::INVALID_CONFIG_PARAMETER);
+        }
         else
-            throw Exception("Unknown handler type '" + handler_type +"' in config here: " +
-                            prefix + "." + key + ".handler.type",ErrorCodes::INVALID_CONFIG_PARAMETER);
+            throw Exception("Unknown element in config: " + prefix + "." + key + ", must be 'rule' or 'defaults'",
+                ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
     }
 
     return main_handler_factory.release();
 }
 
-static const auto ping_response_expression = "Ok.\n";
-static const auto root_response_expression = "config://http_server_default_response";
-
-static inline Poco::Net::HTTPRequestHandlerFactory * createHTTPHandlerFactory(
-    IServer & server, const std::string & name, AsynchronousMetrics & async_metrics)
+static inline Poco::Net::HTTPRequestHandlerFactory * createHTTPHandlerFactory(IServer & server, const std::string & name, AsynchronousMetrics & async_metrics)
 {
     if (server.config().has("http_handlers"))
-        return createHandlersFactoryFromConfig(server, name, "http_handlers");
+        return createHandlersFactoryFromConfig(server, name, "http_handlers", async_metrics);
     else
     {
         auto factory = std::make_unique<HTTPRequestHandlerFactoryMain>(name);
-
-        auto root_handler = std::make_unique<HandlingRuleHTTPHandlerFactory<StaticRequestHandler>>(server, root_response_expression);
-        root_handler->attachStrictPath("/")->allowGetAndHeadRequest();
-        factory->addHandler(root_handler.release());
-
-        auto ping_handler = std::make_unique<HandlingRuleHTTPHandlerFactory<StaticRequestHandler>>(server, ping_response_expression);
-        ping_handler->attachStrictPath("/ping")->allowGetAndHeadRequest();
-        factory->addHandler(ping_handler.release());
-
-        auto replicas_status_handler = std::make_unique<HandlingRuleHTTPHandlerFactory<ReplicasStatusHandler>>(server);
-        replicas_status_handler->attachNonStrictPath("/replicas_status")->allowGetAndHeadRequest();
-        factory->addHandler(replicas_status_handler.release());
+        addDefaultHandlersFactory(*factory, server, &async_metrics);
 
         auto query_handler = std::make_unique<HandlingRuleHTTPHandlerFactory<DynamicQueryHandler>>(server, "query");
```

Comment:
User may expect that `<defaults/>` configures all default handlers including `/query`. I'm not sure, but may be we should add it too? If user wants to disable one of default handlers it's still possible to configure others manually.

Reply:
I didn't look at the code and yes, I expect that `defaults` will enable all handlers that are enabled when there is no customization.